File: core/src/main/java/io/outright/xj/core/model/event/Event.java
// Copyright Outright Mental, Inc. All Rights Reserved.
package io.outright.xj.core.model.event;

import io.outright.xj.core.primitive.beat.Beat;
import io.outright.xj.core.primitive.inflection.Inflection;
import io.outright.xj.core.primitive.note.Note;
import io.outright.xj.core.primitive.tonality.Tonality;
import io.outright.xj.core.primitive.velocity.Velocity;

public interface Event {
  /**
   * Event Velocity (ratio)
   *
   * @return Velocity
   */
  Velocity Velocity();

  /**
   * Event Tonality (ratio)
   *
   * @return Tonality
   */
  Tonality Tonality();

  /**
   * Event Tonality (ratio)
   *
   * @return Inflection
   */
  Inflection Inflection();

  /**
   * Event Position (beats)
   *
   * @return Beat
   */
  Beat Position();

  /**
   * Event Duration (beats)
   *
   * @return Beat
   */
  Beat Duration();

  /**
   * Event Note (music theory)
   *
   * @return Note
   */
  Note Note();
}

File: ui/hub-ui/app/models/event.js
//  Copyright (c) 2018, XJ Music Inc. (https://xj.io) All Rights Reserved.
import Model, {attr, belongsTo} from '@ember-data/model';
import {computed} from '@ember/object';

export default Model.extend({
  pattern: belongsTo({}),
  track: belongsTo({}),
  duration: attr('number'),
  note: attr('string'),
  position: attr('number'),
  velocity: attr('number'),

  title: computed('note', 'position', function () {
    return `${this.note}@${this.position}`;
  }),

});
